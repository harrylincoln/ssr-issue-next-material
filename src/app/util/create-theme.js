import { createMuiTheme } from '@material-ui/core/styles';

const breakpointValues = {
    xs: 0,
    sm: 576,
    md: 768,
    lg: 992,
    xl: 1200,
};

export function themeCreator(primary = {}) {
    return createMuiTheme({
        breakpoints: { values: breakpointValues },
        typography: {
            useNextVariants: true,
        },
        pallete: {
            ...primary,
        },
    });
}
