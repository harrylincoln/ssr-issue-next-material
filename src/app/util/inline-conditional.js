export default (condition, truthy, falsy = null) => {
    if (condition) return [truthy];
    if (falsy === null) return [];
    return [falsy];
};
