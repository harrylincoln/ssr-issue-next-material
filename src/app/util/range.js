export default (length, offset = 0) =>
    Array(length)
        .fill(undefined)
        .map((_, i) => offset + i);
