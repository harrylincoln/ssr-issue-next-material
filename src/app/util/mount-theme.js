import { JssProvider, SheetsRegistry, ThemeProvider } from 'react-jss';
import { mount, shallow } from 'enzyme';

import { Provider } from 'react-redux';
import React from 'react';
import configureMockStore from 'redux-mock-store';
// theming
import primary from '../styles/themes/primary';
import { themeCreator } from './create-theme';
import thunk from 'redux-thunk';
import { createGenerateClassName } from '@material-ui/core/styles';

const mockStore = configureMockStore([thunk]);
const primaryTheme = themeCreator({ primary });

/**
 * JSS helper mount
 * Wraps the passed in component in Themeprovider and captures the styles generated
 * @param {React component} Component
 * @returns {Array} First item in array is the stylesheet generated by jss, the second is the rendered component for further expects
 */
export const themeMount = (Component, store = mockStore({})) => {
    const sheets = new SheetsRegistry();
    const ClonedComponent = React.cloneElement(Component, {
        _renderid: 'rendered-component' /* eslint-disable-line */,
    });

    const fullTree = mount(
        <JssProvider
            generateClassName={createGenerateClassName()}
            registry={sheets}
        >
            <ThemeProvider theme={primaryTheme}>
                <Provider store={store}>{ClonedComponent}</Provider>
            </ThemeProvider>
        </JssProvider>
    );

    const mountWithTheme = (tree, theme) => {
        const context = shallow(<ThemeProvider theme={theme} />, {
            lifecycleExperimental: true,
        })
            .instance()
            .getChildContext();

        return mount(tree, {
            context,
            childContextTypes: ThemeProvider.childContextTypes, // Needed so child components receive the theme prop
        });
    };

    const fullMountWithTheme = mountWithTheme(
        <Provider store={store}>
            <JssProvider
                generateClassName={createGenerateClassName()}
                registry={sheets}
            >
                {Component}
            </JssProvider>
        </Provider>,
        primaryTheme
    );

    return {
        styleSheets: sheets.toString(),
        component: fullTree.findWhere(
            (node) =>
                node.props()._renderid === /* eslint-disable-line */
                'rendered-component'
        ),
        fullRendered: fullTree,
        fullMountWithTheme,
        mockStore: store,
    };
};
