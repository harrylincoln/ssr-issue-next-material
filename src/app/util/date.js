import range from './range';

export const getMidnight = (date = new Date()) => {
    const clone = new Date(date);
    clone.setHours(0);
    clone.setMinutes(0);
    clone.setSeconds(0);
    clone.setMilliseconds(0);

    return clone;
};

export const getNextSaturday = (date) => {
    const currentDay = date.getDay();

    const daysUntilSaturday = currentDay === 6 ? 6 - currentDay : 7;

    const saturday = new Date(date);

    saturday.setDate(date.getDate() + daysUntilSaturday);

    return saturday;
};

export const addWeek = (date, weeks = 1) => {
    const clone = new Date(date);
    clone.setDate(date.getDate() + 7 * weeks);
    return clone;
};

export const getWeeks = (startDate, amount) =>
    range(amount, 1).map((week) => addWeek(startDate, week));

export const months = [
    'January',
    'February',
    'March',
    'April',
    'May',
    'June',
    'July',
    'August',
    'September',
    'October',
    'November',
    'December',
];

export const getOrdinal = (day) => {
    const ordinals = ['', 'st', 'nd', 'rd'];
    const remainder = day % 100;
    return (
        day +
        (remainder > 10 && remainder < 14
            ? 'th'
            : ordinals[remainder % 10] || 'th')
    );
};
