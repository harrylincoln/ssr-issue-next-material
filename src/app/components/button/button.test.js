import Button from './button';
import Link from 'next/link';
import React from 'react';
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { themeMount } from '../../util/mount-theme';
import Router from 'next/router';

const mockStore = configureStore([thunk]);
const store = mockStore({
    countries: {
        countries: [],
        currentCountry: {
            name: 'Narnia',
            iso2: 'NA',
        },
    },
});

const mockedRouter = { push: () => {}, prefetch: () => {} };
Router.router = mockedRouter;

describe('Button', () => {
    it('should render correctly', () => {
        const { styleSheets, component } = themeMount(
            <Button>Test</Button>,
            store
        );

        expect(styleSheets.toString()).toMatchSnapshot();
        expect(component).toMatchSnapshot();
    });

    it('should render an anchor tag if URL prop is set', () => {
        const { component } = themeMount(
            <Button url="http://google.com">Test</Button>
        );
        expect(
            component
                .find('Button')
                .children()
                .type()
        ).toEqual('a');
    });

    it('should render a button tag if NO URL prop is set', () => {
        const { component } = themeMount(<Button>Test</Button>);
        expect(
            component
                .find('Button')
                .children()
                .type()
        ).toEqual('button');
    });

    // it('should apply onClick event to child', () => {});

    it('should wrap the button in a <Link> component if route prop is set', () => {
        const { component } = themeMount(
            <Button route="/path/toNav">Test</Button>
        );
        expect(
            component
                .find('WithStyles(Button)')
                .find('Button')
                .children()
                .type()
        ).toEqual(Link);
        expect(
            component
                .find('WithStyles(Button)')
                .find('Button')
                .children()
                .instance().props.passHref
        ).toBe(true);
    });

    it('should prefetch props to <Link> component if they are set', () => {
        const { component } = themeMount(
            <Button route="/path/toNav" prefetch>
                Test
            </Button>
        );

        expect(
            component
                .find('WithStyles(Button)')
                .find('Button')
                .children()
                .type()
        ).toEqual(Link);
        expect(
            component
                .find('WithStyles(Button)')
                .find('Button')
                .children()
                .instance().props.prefetch
        ).toBe(true);
    });

    it('should apply prettyUrl to href if "as" prop is set', () => {
        const { component } = themeMount(
            <Button route="/path/toNav" as="/prettyUrl/isnt/it/nice">
                Test
            </Button>
        );

        expect(
            component
                .find('WithStyles(Button)')
                .find('Button')
                .children()
                .type()
        ).toEqual(Link);
        expect(
            component
                .find('WithStyles(Button)')
                .find('Button')
                .children()
                .instance().props.as
        ).toEqual('/prettyUrl/isnt/it/nice');
    });
});

/*

Tests to add

onClick being applied to child
route / as / prefetch ~ add props / render Link

*/
