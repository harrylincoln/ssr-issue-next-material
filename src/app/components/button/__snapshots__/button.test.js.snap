// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button should render correctly 1`] = `
".button-test {
  color: #FFFFFF;
  cursor: pointer;
  display: inline-block;
  padding: 1rem 3.2rem;
  font-size: 1.6rem;
  line-height: 2rem;
  user-select: none;
  font-weight: 500;
  border-style: solid;
  border-width: 1px;
  border-radius: .4rem;
  background-color: #3E51B5;
}
.button-test:hover {
  background: #0F287F;
}
.button-test:focus {
  outline: none;
  box-shadow: 0 0 0 2px #E2198B;
  border-radius: 4px;
  background-color: #3E51B5;
}
.button-test:disabled {
  color: #121212;
  cursor: auto;
  background: #F2F2F2;
}
.button-test svg {
  margin: 0 0.8rem 0 0;
  display: inline-block;
  vertical-align: text-bottom;
}
.small-test {
  padding: .4rem 1.8rem;
}
.secondary-test {
  color: #3E51B5;
  padding: 0.5rem 1rem;
  background: #DADADA;
  border-color: #595959;
}
.secondary-test:hover {
  color: white;
  background-color: #3E51B5;
}
.tertiary-test {
  color: #3E51B5;
  background: rgba(62,81,181, .05);
  border-color: rgba(62,81,181, .25);
}
.tertiary-test:hover {
  color: white;
  background: rgba(62,81,181, .75);
}
.tertiary-test:focus {
  color: white;
  background: rgba(62,81,181, .75);
}
.light-test {
  color: #FFFFFF;
  background: rgba(255,255,255, .05);
  border-color: rgba(255,255,255, .25);
}
.light-test:hover {
  color: inherit;
  background: rgba(255,255,255, .75);
}
.edit-test {
  color: #3E51B5;
  padding: 0.5rem 1rem;
  background-color: white;
}
.edit-test:before {
  width: 25px;
  height: 13px;
  display: inline-block;
  content: '';
  background-size: 26px 13px;
  background-image: url('/static/edit.svg');
}
.edit-test:hover {
  color: white;
}
.edit-test:hover:before {
  background-image: url('/static/edit-white.svg');
}
.full-test {
  width: 100%;
  max-width: 500px;
  text-align: center;
  padding-left: 1rem;
  padding-right: 1rem;
}
.selected-test {
  color: #FFFFFF;
  cursor: pointer;
  display: inline-block;
  padding: 1rem 3.2rem;
  font-size: 1.6rem;
  line-height: 2rem;
  user-select: none;
  font-weight: 500;
  border-style: solid;
  border-width: 1px;
  border-radius: .4rem;
  background-color: #E3120B;
}
.selected-test:hover {
  background-color: #E3120B;
}
.selected-test:focus {
  outline: none;
  box-shadow: 0 0 0 2px #E2198B;
  border-radius: 4px;
  background-color: #E3120B;
}
.full_width_bg-test {
  margin: -12px 0;
  padding: 12px;
  background: #E3120B;
}
.dark-test {
  color: #FFFFFF;
  border-color: #383E42;
  background-color: #383E42;
}
.dark-test:hover {
  background-color: #121212;
}
.dark-test:focus {
  border-color: #16C9B3;
  background-color: #121212;
}
.dark-test:active {
  border-color: #16C9B3;
  background-color: #121212;
}"
`;

exports[`Button should render correctly 2`] = `
Array [
  <WithStyles(Button)
    _renderid="rendered-component"
  >
    <Button
      _renderid="rendered-component"
      as={null}
      classes={
        Object {
          "button": "button-test",
          "dark": "dark-test",
          "edit": "edit-test",
          "full": "full-test",
          "full_width_bg": "full_width_bg-test",
          "light": "light-test",
          "secondary": "secondary-test",
          "selected": "selected-test",
          "small": "small-test",
          "tertiary": "tertiary-test",
        }
      }
      dark={false}
      disabled={false}
      full={false}
      id={null}
      light={false}
      onClick={null}
      prefetch={false}
      route={null}
      secondary={false}
      selected={false}
      small={false}
      tabIndex={null}
      tertiary={false}
      text=""
      type={null}
      url={null}
    >
      <button
        _renderid="rendered-component"
        className="button-test"
        disabled={false}
        id={null}
        tabIndex={null}
        type={null}
      >
        Test
      </button>
    </Button>
  </WithStyles(Button)>,
  <Button
    _renderid="rendered-component"
    as={null}
    classes={
      Object {
        "button": "button-test",
        "dark": "dark-test",
        "edit": "edit-test",
        "full": "full-test",
        "full_width_bg": "full_width_bg-test",
        "light": "light-test",
        "secondary": "secondary-test",
        "selected": "selected-test",
        "small": "small-test",
        "tertiary": "tertiary-test",
      }
    }
    dark={false}
    disabled={false}
    full={false}
    id={null}
    light={false}
    onClick={null}
    prefetch={false}
    route={null}
    secondary={false}
    selected={false}
    small={false}
    tabIndex={null}
    tertiary={false}
    text=""
    type={null}
    url={null}
  >
    <button
      _renderid="rendered-component"
      className="button-test"
      disabled={false}
      id={null}
      tabIndex={null}
      type={null}
    >
      Test
    </button>
  </Button>,
  <button
    _renderid="rendered-component"
    className="button-test"
    disabled={false}
    id={null}
    tabIndex={null}
    type={null}
  >
    Test
  </button>,
]
`;
